From 80d0d1a82c11fab59039833fa0a76d9f13ad449e Mon Sep 17 00:00:00 2001
From: Paolo Bonzini <pbonzini@redhat.com>
Date: Tue, 22 Jul 2014 14:37:49 -0400
Subject: [virt] kvm/x86: add comments where MMIO does not return to the emulator

Message-id: <1406040016-3289-75-git-send-email-pbonzini@redhat.com>
Patchwork-id: 86153
O-Subject: [RHEL7 PATCH v2 074/221] KVM: x86: add comments where MMIO does not return to the emulator
Bugzilla: 1116936
RH-Acked-by: Radim Krcmar <rkrcmar@redhat.com>
RH-Acked-by: Bandan Das <bsd@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

Support for single-step in the emulator (new in 3.12) does not work for
MMIO or PIO writes, because they are completed without returning to
the emulator.  This is not worse than what we had in 3.11; still, add
comments so that the issue is not forgotten.

Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>
Signed-off-by: Gleb Natapov <gleb@redhat.com>
(cherry picked from commit 0912c9771e9902f752e890e93af495cc06a786ac)
Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/x86/kvm/x86.c b/arch/x86/kvm/x86.c
index 2327d8e..f0cdd18 100644
--- a/arch/x86/kvm/x86.c
+++ b/arch/x86/kvm/x86.c
@@ -5189,9 +5189,10 @@ restart:
   inject_emulated_exception(vcpu);
   r = EMULATE_DONE;
  } else if (vcpu->arch.pio.count) {
-  if (!vcpu->arch.pio.in)
+  if (!vcpu->arch.pio.in) {
+   /* FIXME: return into emulator if single-stepping.  */
    vcpu->arch.pio.count = 0;
-  else {
+  } else {
    writeback = false;
    vcpu->arch.complete_userspace_io = complete_emulated_pio;
   }
@@ -6188,6 +6189,8 @@ static int complete_emulated_mmio(struct kvm_vcpu *vcpu)
 
  if (vcpu->mmio_cur_fragment >= vcpu->mmio_nr_fragments) {
   vcpu->mmio_needed = 0;
+
+  /* FIXME: return into emulator if single-stepping.  */
   if (vcpu->mmio_is_write)
    return 1;
   vcpu->mmio_read_completed = 1;
-- 
1.7.1